Billing Backend

AbhishekGupta10004 opened the pull request.

Commit:
GET working , POST not going to db

Commit:
GET working post not going to db

Commit:
Backend complete and working

johnstoecker commented on api/src/main/resources/liquibase.xml:
```diff
@@ -515,5 +515,502 @@
+            baseTableName="raxacore_billing"
+            constraintName="FK_raxacore_billing_provider_id_Provider_provider_id"
+            referencedColumnNames="provider_id"
+            referencedTableName="Provider" />
```
provider table name needs to be lower case (windows this is okay, fails in linux)

johnstoecker commented on api/src/main/resources/liquibase.xml:
```diff
@@ -515,5 +515,502 @@
+            baseTableName="raxacore_billing"
+            constraintName="FK_raxacore_billing_patient_id_Patient_patient_id"
+            referencedColumnNames="patient_id"
+            referencedTableName="Patient" />
```
Patient table should be lower case

johnstoecker commented on api/src/main/resources/liquibase.xml:
```diff
@@ -515,5 +515,502 @@
+            baseTableName="raxacore_billing_item"
+            constraintName="FK_raxacore_billing_item_provider_id_Provider_provider_id"
+            referencedColumnNames="provider_id"
+            referencedTableName="Provider" />
```
provider lower case

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/BillingItem.java:
```diff
@@ -0,0 +1,174 @@
+		
+	}
+	
+	public Integer getbillItemId() {
```
should be camelCase, `getBillItemId`.
Inconsitent in class name and method names. BillingItem or BillItem?

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/Billing.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * PatientList stores the data needed to lists patients for registration, screener, etc.
+ */
+public class Billing extends BaseOpenmrsMetadata implements Serializable {
```
be consistent in naming: Bill or Billing?

If Bill: class should be called `Bill` (line 26)
If Billing:  integer id should be called `billingId` (line 28, and update others below like `getBillId`)

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/Billing.java:
```diff
@@ -0,0 +1,104 @@
+	public void setId(Integer arg0) {
+		// TODO Auto-generated method stub
+		
+	}
```
should remove `getBillId` and `setBillId`, and put their functionality into these required override methods.

nathanleiby replied:
Nevermind - appears this is standard in openMRS. You should at least have `setId` call `setBillId` though. (same idea for getId)

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/BillingItemAdjustment.java:
```diff
@@ -0,0 +1,97 @@
+
+public class BillingItemAdjustment extends BaseOpenmrsMetadata implements Serializable {
+	
+	private Integer billItemAdjustmentId;
```
naming consistency -- `billItemAdjustmentId` vs `BillingItemAdjustment`. choose one and be consistent through all classes.

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/db/hibernate/HibernateDrugPurchaseOrderDAO.java:
```diff
@@ -39,7 +39,8 @@ public void setSessionFactory(SessionFactory sessionFactory) {
 	public DrugPurchaseOrder saveDrugPurchaseOrder(DrugPurchaseOrder drugPurchaseOrder) {
 		
-		sessionFactory.getCurrentSession().saveOrUpdate(drugPurchaseOrder);
+		//sessionFactory.getCurrentSession().saveOrUpdate(drugPurchaseOrder);  //did not work 
```
remove this line if doesn't work. 

What was the bug that was happening before? Can you write a test to verify that the new code does work?

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/impl/BillingItemServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+	public void onShutdown() {
+		
+		log.info("Starting drug purchase order service");
+	}
```
text is wrong. 
- not a drug purchase order - this is a billing item
- also, this is the `onShutdown` method, not the startup method

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/impl/BillingItemServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+	
+	public void onStartup() {
+		
+		log.info("Starting drug purchase order service");
```
text is wrong. 
- not a drug purchase order - this is a billing item

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/impl/BillingServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+	
+	public void onShutdown() {
+		
+		log.info("Starting drug purchase order service");
```
text is wrong. 
- not a drug purchase order - this is Billing
- also, this is the `onShutdown` method, not the startup method

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/impl/BillingServiceImpl.java:
```diff
@@ -0,0 +1,80 @@
+	
+	public void onStartup() {
+		
+		log.info("Starting drug purchase order service");
```
text is wrong. 
- not a drug purchase order - this is Billing

nathanleiby commented on api/src/test/java/org/raxa/module/raxacore/db/hibernate/HibernateDrugInventoryDAOTest.java:
```diff
@@ -99,17 +98,17 @@ public void testDeleteDrugInventory() {
 		
-		DrugInventory result=dao.getDrugInventoryByUuid("68547121-1b70-465c-99ee-c9dfd95e7d34");
-		assertEquals(result,null);
+		DrugInventory result = dao.getDrugInventoryByUuid("68547121-1b70-465c-99ee-c9dfd95e7d34");
```
This is a "magic number" (well, string) in that you use it frequently in many places without explaining it . See "unnamed numerical constants" on the Wiki page for a justification of why not to do this. The UUID `"68547121-1b70-465c-99ee-c9dfd95e7d34"` appears on at least one line and the UUID "68547121-1b70-465c-99ee-c9dfd95e7d36" appears on at least 3 lines. 
http://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants

Instead, you should abstract it to a shared constant so you only have to update it in one place, and to clarify the functionality of the UUID.

above:
`string DRUG_INVENTORY_UUID = "68547121-1b70-465c-99ee-c9dfd95e7d34";`
`string SOMETHING_ELSE = "68547121-1b70-465c-99ee-c9dfd95e7d36";`
here and elsewhere:
`DrugInventory result = dao.getDrugInventoryByUuid(DRUG_INVENTORY_UUID);`
etc

nathanleiby replied:
also, is this UUID server specific? I'm not yet familiar with how this is handled on back-end and if it's ok to hard-core a UUID like this, but on front-end it's a problem because UUID's are different depending on the OpenMRS server.

nathanleiby replied:
Whoops, I see this is a test. Will it work across all computers/servers despite the UUID thing I mentioned?

nathanleiby commented on omod/src/main/java/org/raxa/module/raxacore/web/v1_0/controller/DrugInventoryController.java:
```diff
@@ -93,14 +100,25 @@ public Object saveDrugInventory(@RequestBody SimpleObject post, HttpServletReque
+			obj.add("quantity", created.getQuantity());
+		}
+		catch (Exception e) {
+			System.out.println("helllloooooo errroorr ocuuured");
```
would be helpful to put a more specific error message, if possible :) e.g. "Failed to save DrugInventory via <X> service"

nathanleiby commented on omod/src/main/java/org/raxa/module/raxacore/web/v1_0/controller/DrugInventoryController.java:
```diff
@@ -93,14 +100,25 @@ public Object saveDrugInventory(@RequestBody SimpleObject post, HttpServletReque
+		}
+		DrugInventory created;
+		SimpleObject obj = obj = new SimpleObject();
+		;
```
You have some unneeded code here. Simply
`SimpleObject obj = new SimpleObject();`
would be sufficient

nathanleiby commented on omod/src/main/java/org/raxa/module/raxacore/web/v1_0/controller/DrugPurchaseOrderController.java:
```diff
@@ -65,32 +67,42 @@ private String getResourceVersion() {
-		drugOrder.setUuid(post.get("uuid").toString());
 		drugOrder.setName(post.get("name").toString());
+		drugOrder.setProvider(Context.getProviderService().getProvider(Integer.parseInt(post.get("providerId").toString())));
+		//drugOrder.setProviderId(new Integer(1));
```
please remove commented code, if not used

nathanleiby commented on omod/src/main/java/org/raxa/module/raxacore/web/v1_0/controller/DrugInventoryController.java:
```diff
@@ -93,14 +100,25 @@ public Object saveDrugInventory(@RequestBody SimpleObject post, HttpServletReque
+		if (post.get("drugPurchaseOrderId") != null) {
+			drugInventory.setDrugPurchaseOrderId(Integer.parseInt(post.get("drugPurchaseOrderId").toString()));
+		}
+		DrugInventory created;
```
could increase readability by writing 
`DrugInventory created = service.saveDrugInventory(drugInventory);`
instead of instantiating the variable above and then saving a value to it below.
(move all of this within the try block)

nathanleiby commented on omod/src/main/java/org/raxa/module/raxacore/web/v1_0/controller/DrugPurchaseOrderController.java:
```diff
@@ -65,32 +67,42 @@ private String getResourceVersion() {
-		}
+		drugOrder.setReceived(false);
+		
+		//
```
give a quick look to formatting and cleanup in your files. Try to remove things like unneeded comments, printlns, etc.

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/impl/BillingItemAdjustmentServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+	
+	public void onShutdown() {
+		
+		log.info("Starting drug purchase order service");
```
text is wrong. 
- not a drug purchase order - this is a BillItemAdjustment
- also, this is the `onShutdown` method, not the startup method

nathanleiby commented on api/src/main/java/org/raxa/module/raxacore/impl/BillingItemAdjustmentServiceImpl.java:
```diff
@@ -0,0 +1,78 @@
+	
+	public void onStartup() {
+		
+		log.info("Starting drug purchase order service");
```
text is wrong. 
- not a drug purchase order - this is a BillItemAdjustment

johnstoecker commented on api/src/main/java/org/raxa/module/raxacore/Billing.java:
```diff
@@ -0,0 +1,104 @@
+ */
+public class Billing extends BaseOpenmrsMetadata implements Serializable {
+	
+	private Integer billId;
```
Change all ids to private Integer id

johnstoecker commented on api/src/main/java/org/raxa/module/raxacore/BillingItem.java:
```diff
@@ -0,0 +1,174 @@
+ */
+public class BillingItem extends BaseOpenmrsMetadata implements Serializable {
+	
+	private Integer billItemId;
```
change to id

nathanleiby commented:
@AbhishekGupta10004 how is this coming along?

Commit:
This the biiling backend with tests and Rest services.
More rest calls at the backend will be made as per the requirement in
the front end.

johnstoecker commented:
Closing for new pull request
